File: bench/spectralnorm/spectralnorm.lua
-- The Computer Language Shootout
-- http://shootout.alioth.debian.org/
-- contributed by Isaac Gouy
-- modified by Adrian Sietsma


function A(i,j) local x = i+j;return 2/(x*x+x  +i+i+2) end

function Av(n,v)
   local Av = {}
   for i = 0,n-1 do
      local a = A(i,0)*v[0]
      for j = 1,n-1 do a = a + (A(i,j)*v[j]) end
      Av[i] = a
   end
   return Av
end

function Atv(n,v)
   local Atv = {}
   for i = 0,n-1 do
      local a = A(0,i)*v[0]
      for j = 1,n-1 do a = a + A(j,i)*v[j] end
      Atv[i] = a
   end
   return Atv
end

function AtAv(n,v)
   return Atv(n, Av(n,v))
end


local n = tonumber(arg and arg[1]) or 100

local u,v = {},{}
for i = 0,n-1 do u[i] = 1 end

for i = 1,10 do v = AtAv(n,u); u = AtAv(n,v) end

local vi = v[0]
local vBv,vv = u[0]*vi, vi*vi
for i = 1,n-1 do vi = v[i];vBv = vBv + u[i]*vi; vv = vv + vi*vi end

io.write( string.format("%0.9f", math.sqrt(vBv/vv) ), "\n")

File: bench/spectralnorm/spectralnorm.scala
/* The Computer Language Benchmarks Game
   http://shootout.alioth.debian.org/
   contributed by Isaac Gouy
   modified by Meiko Rachimow
*/

object spectralnorm {

   def main(args: Array[String]) = {
      val n = {if (args.length > 0) Integer.parseInt(args(0)) else 100}
      val formatter = new java.text.DecimalFormat("#.000000000");
      println(formatter.format(new SpectralNorm() approximate(n)))
   }
}

class SpectralNorm(){

   type Vector = Array[double]

   def approximate(n: int)= {
      val u = new Vector(n)
      var i = 0;
      while (i < n){ u(i) = 1.0; i = i+1 }

      val v = new Vector(n)

      i = 0
      while (i < 10){
         multiplyAtAv(n,u,v)
         multiplyAtAv(n,v,u)
         i = i+1
      }

      var vbv = 0.0; var vv = 0.0
      i = 0
      while (i < n){
         vbv = vbv + u(i)*v(i)
         vv = vv + v(i)*v(i)
         i = i+1
      }

      Math sqrt(vbv/vv)
   }


   def a(i: int, j: int) = 1.0/((i+j)*(i+j+1)/2 +i+1)

   def multiplyAv(n: int, v: Vector, av: Vector) = {
      var i = 0
      while (i < n){
         av(i) = 0.0
         var j = 0
         while (j < n){ av(i) = av(i) + a(i,j) * v(j); j = j+1 }
         i = i+1
      }
   }

   def multiplyAtv(n: int, v: Vector, atv: Vector) = {
      var i = 0
      while (i < n){
         atv(i) = 0.0
         var j = 0
         while (j < n){ atv(i) = atv(i) + a(j,i) * v(j); j = j+1 }
         i = i+1
      }
   }

   def multiplyAtAv(n: int, v: Vector, atav: Vector) = {
      val u = new Vector(n)
      multiplyAv(n,v,u)
      multiplyAtv(n,u,atav)
   }

}
